Add YAML gas reactions (attempt 2, lite-version)

Partmedia opened the pull request:
## About the PR
This adds YAML gas reactions, i.e. gas reactions that can be defined in YAML.

This replaces a bunch of [hard coded C# reaction code with boilerplate](https://github.com/space-wizards/space-station-14/blob/6869adf/Content.Server/Atmos/Reactions/AmmoniaOxygenReaction.cs) with something like this:

```yaml
- type: gasReaction
  id: AmmoniaOxygenReaction
  minimumTemperature: 323.149
  reactants:
    Ammonia: 2
    Oxygen: 2
  products:
    NitrousOxide: 1
    WaterVapor: 3
```

This is attempt 2 after https://github.com/space-wizards/space-station-14/pull/22803

Since the original, there have been additional guards against divide by zero and fixes elsewhere in atmos code.

We also reduce the diff to make this more reviewable by including only the ammonia + oxygen reaction and the N2O decomposition reaction. We can slowly add the other reactions once we determine there are no bugs here.

## Why / Balance
People kept copy/pasting gas reactions to create new ones. However, copy/pasted reactions:

1. Require code changes.
2. Lead to spaghetti copy/pasted code.
3. Do not model concentration-dependent reaction effects.
4. Do not model temperature-dependent reaction effects.
5. Do not model catalyst effects.
6. Are not conducive to future code optimizations.

## Technical details
A lot of the code required to implement a gas reaction can actually be shared by multiple reactions. For each reaction, this code:

1. Loops through all reactants, computing the reaction rate to model concentration-dependent effects. This means that not removing products or waste products will actually "soft clog" the reaction.
2. Loops through all catalysts, modifying the reaction activation energy depending on the presence of the catalyst.
3. Applies temperature-dependent rate scaling. Currently, many reactions do not scale with temperature, or only scale linearly with temperature. This code uses the [Arrhenius equation](https://en.wikipedia.org/wiki/Arrhenius_equation) to model exponential increase in reaction rate depending on temperature, while accounting for presence of catalysts.
4. Goes back through the reactants, removing the reacted parts according to the stoichiometry of the equation.
5. Goes goes through the products, adding the products according to the stoichiometry of the equation.
6. Adds or removes the energy released/absorbed by the reaction, defined by the YAML `enthalpy` field.
7. Adds any fire/hotspots as needed.

This code accomplishes all this while preserving the existing API for any existing reactions. Reactions can still continue to use custom reaction code defined in C# by not defining the `reactants` field.

This represents a "naive" implementation of a generic reaction. In the future, this can be improved with a simultaneous (non-priority based) matrix-based SIMD-accelerated solver.

## Media
https://github.com/space-wizards/space-station-14/assets/3229565/6fd60977-e213-4dad-be1c-8f844d30b6c8

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
None, but an older version of this code caused strange atmos bugs.

![image](https://github.com/space-wizards/space-station-14/assets/3229565/0f5a4ebd-e3d3-4fed-a9db-3cf3e4e2e9df)

Please help review and test this code. Don't merge until at least 2024-06-20 at which point I'll have some time to monitor the servers in case this Heisenbug comes back.

**Changelog**
:cl: notafet
- tweak: N2O synthesis and decomposition now has a concentration-dependent reaction rate.

Partmedia committed:
Add YAML gas reactions

Partmedia committed:
Ammonia reaction

Partmedia committed:
N2O decomposition

slarticodefast commented on Content.Server/Atmos/EntitySystems/GenericGasReactionSystem.cs:
```diff
@@ -0,0 +1,115 @@
+    {
+        float nTotal = mix.TotalMoles;
+        // Guard against very small amounts of gas in mixture
+        if (nTotal < Atmospherics.GasMinMoles)
```
This should be checked for each of the reaction ingredients individually at some point. For example, you don't want to do the N2O Decomposition calculation in pure oxygen. Also, GasMinMoles = 0.00000005f is a really low threshold and you can probably keep the old ones of 0.01 moles per gas. If you have a big room with a tiny trace amount of a gas, you don't want to do the reaction on every single tile. This should ensure better numerical stability as well.

Partmedia replied:
This suggests to me that `GasMinMoles` should perhaps be raised in a separate PR. Or perhaps we should create a separate global set to 0.01 and make this a global minimum reaction amount for all reactions? It doesn't seem to make sense to me to allow different minimums for different reactions.

slarticodefast commented on Content.Server/Atmos/EntitySystems/GenericGasReactionSystem.cs:
```diff
@@ -0,0 +1,115 @@
+            float rate = 1f; // rate of this reaction
+            foreach (var (reactant, num) in reaction.Reactants)
+            {
+                float concentration = mix.GetMoles(reactant)/nTotal;
```
```suggestion
                float concentration = mix.GetMoles(reactant)/mix.Volume;
```
The concentration should be in moles/volume, not the mole fraction. If you have the same gas amount in a bigger container, it should react slower since the molecules are farther apart and less likely to hit each other. The rates might need rebalancing for this.

Partmedia replied:
Possible divide by near-zero, perhaps multiplying by pressure would be better.

slarticodefast commented on Content.Server/Atmos/EntitySystems/GenericGasReactionSystem.cs:
```diff
@@ -0,0 +1,115 @@
+            float catalystEnergy = 0;
+            foreach (var (catalyst, dE) in reaction.Catalysts)
+            {
+                float concentration = mix.GetMoles(catalyst)/nTotal;
```
same thing

Partmedia replied:
Ditto.

slarticodefast reviewed (commented):
I really like the idea of this. The implementation seems pretty straightforward as well. I commented on a few things I noticed so far. I will do some more in-game testing for numerical stability in the next few days.

CBozo commented:
Pardon my stupidity if this is a dumb question, but how would this work for gasses like plasma? I know that it's combustion ratio changes based on how hot it is, so would it stay hard coded or would it be implemented it YAML form?

Partmedia commented on Content.Server/Atmos/EntitySystems/GenericGasReactionSystem.cs:
```diff
@@ -0,0 +1,115 @@
+        if (temp < Atmospherics.TCMB)
+            return 0;
+
+        return A*MathF.Exp(-Ea/(Atmospherics.R*temp));
```
While it is mathematically equivalent, this suggestion sounds like premature optimization. Note that both `RateScaleFactor` and `TScaleFactor` are constants, and so `A` is also a constant. Multiplying a float by a constant is quite cheap. Combining it also makes the code harder to understand.

Partmedia replied:
To add to this, making the intent clear is important for a future SIMD version, where multiply-by-const can be done SIMD.

Partmedia commented on Content.Server/Atmos/EntitySystems/AtmosphereSystem.Gases.cs:
```diff
@@ -313,6 +314,7 @@ public GasCompareResult CompareExchange(GasMixture sample, GasMixture otherSampl
+            // First pass: run through the legacy (hard-coded) gas reactions
             var reaction = ReactionResult.NoReaction;
             var temperature = mixture.Temperature;
             var energy = GetThermalEnergy(mixture);
```
I've implemented this suggestion.

TheShuEd commented:
@Partmedia @metalgearsloth 
What the status of this PR?
Did it derelict?

Partmedia commented:
I think this would still be an interesting addition to the game, but I need help testing this. And unfortunately I have been focusing on other priorities, so progress on this has been slow. But the code still remains mergable, and if @slarticodefast is interested and would be able to take over steering this PR to make it in, then I think there's still a path forward.

IProduceWidgets commented on Content.Server/Atmos/Reactions/GasReactionPrototype.cs:
```diff
@@ -48,6 +48,27 @@ public sealed partial class GasReactionPrototype : IPrototype
         /// </summary>
         [DataField("effects")] private List<IGasReactionEffect> _effects = new();
 
+        [DataField("enthalpy")]
+        public float Enthalpy;
```
Also I think you dont need the name strings, but Im not 100% sure in a prototype instead of a component.

IProduceWidgets reviewed (changes requested):
Just a couple hygiene things. Nothing jumped out at me codewise, but I haven't pulled it locally yet.

Errant-4 commented:
I merged master as a test, and there is a problem that breaks this, without showing up as a conflict. 
#34967 has turned GasMixture.ReactionResults into a float array, which breaks GenericGasReactionSystem.cs line 101 `mix.ReactionResults[GasReaction.Fire] += rate;`

ArtisticRoomba committed:
Merge branch 'master' into fork/Partmedia/notafet/gas_reactions_lite

# Conflicts:
#	Resources/Prototypes/Atmospherics/reactions.yml

ArtisticRoomba committed:
cleanup merge and fix broken behavior from merge

ArtisticRoomba commented:
Fixed the broken merge. I'll spend a bit testing this, ideally for testing on master I'd probably want to:
- Merge this in the beginning of a release cycle
- Create and merge the PR moving over the rest of the reactions.
- Ask people if they've noticed any odd behavior or differences at the end of stable release.

ArtisticRoomba committed:
Assert that volume is above a min

ArtisticRoomba committed:
Warnings and suggestions cleanup

ArtisticRoomba committed:
Move fix/0 guard further up

ArtisticRoomba committed:
Merge remote-tracking branch 'upstream/master' into fork/Partmedia/notafet/gas_reactions_lite

ArtisticRoomba committed:
Merge branch 'fork/Partmedia/notafet/gas_reactions_lite' into engi-atmos/YAML-gas-reactions

ArtisticRoomba committed:
remove gas reactions benchmark for generics

ArtisticRoomba committed:
Determine concentration based on the mole fraction instead of mol/volume
Slightly unrealistic however things don't work due to near-zero reaction rates and I'd rather get this into a state of testing than being truly 100% accurate

ArtisticRoomba committed:
Migrate over FrezonCoolant and TritiumFire to YAML reactions

ArtisticRoomba committed:
Migrate PlasmaFire and FrezonProduction to generics

ArtisticRoomba committed:
Fix ID and priority

ArtisticRoomba committed:
Add a link to a basic desmos calculator that can help visualize ReactionRate

ArtisticRoomba committed:
Leave comment on reactions occcuring below MinimumTemperatureRequirement for playtest

ArtisticRoomba committed:
Add guard for reactions below MinTemp
This is causing too many problems for balancing gameplay and should be looked at later in further detail. I just want to get things ready for playtest at this point.

ArtisticRoomba committed:
Rebalance PlasmaFireLowEnergy/HighEnergy for TEG

ArtisticRoomba committed:
Revert "Add guard for reactions below MinTemp"

This reverts commit 80f59cabde8029f3e9319a2ff9d47cbe0e413689.

Partmedia committed:
Add mass conservation convergence check

ArtisticRoomba committed:
Merge remote-tracking branch 'upstream/engi-atmos/YAML-gas-reactions' into engi-atmos/YAML-gas-reactions

ArtisticRoomba committed:
Cleanup allocs and naming

ArtisticRoomba committed:
Revert "Cleanup allocs and naming"

This reverts commit 014e3be8d4a623a4a3e447aa580204657be407dc.

ArtisticRoomba committed:
cleanup and fix logger warning

ArtisticRoomba committed:
add maxtemp clamping, extent of reaction fallback solver

Partmedia committed:
Move magic numbers to the top of file

Partmedia committed:
Adjust Ea

Partmedia committed:
Implement LimExp

ArtisticRoomba committed:
merge upstream

ArtisticRoomba committed:
Revert "Migrate PlasmaFire and FrezonProduction to generics"

This reverts commit ae6ffbbe

ArtisticRoomba committed:
add TODO
